### PR TITLE
Add tests to verify flow is rerouted when switch or ISL fails

### LINKED ISSUE #1467: Replace ATDD tests from FlowReinstall.feature with functional tests
Replace all scenarios from https://github.com/telstra/open-kilda/blob/develop/services/src/atdd/src/test/resources/features/mvp.1/FlowReinstall.feature with the corresponding tests in the functional tests module.

Scenarios:
- [x] Re-installing Ingress and Egress flows
- [x] Re-installing transit flow when no other path is available
- [ ] Re-installing transit flow when no other path is available and reflow feature is turned off
- [ ] Re-installing transit flow after changing path

### PR DESCRIPTION
**Pull request by ylobankov**
Closes #1467.

### CODE REVIEW

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy [hunk 1]
```diff
@@ -237,6 +313,54 @@ class AutoRerouteSpec extends BaseSpecification {
         }
     }
 
+    def "Flow in 'Down' status tries to reroute when connecting a new switch"() {
+        requireProfiles("virtual")
+
+        given: "With-intermediate-switch flow with one alternate path at least"
+        def flow = withIntermediateSwitchFlow(true)
+        def allPaths = db.getPaths(flow.source.datapath, flow.destination.datapath)*.path.sort { it.size() }
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy]: This is a heavy operation which is performed twice (first time inside withIntermediateSwitchFlow)

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy [hunk 2]
```diff
@@ -237,6 +313,54 @@ class AutoRerouteSpec extends BaseSpecification {
...
+        }
+
+        then: "Flow goes to 'Up' status"
+        Wrappers.wait(rerouteDelay + discoveryInterval + WAIT_OFFSET) {
+            northboundService.getFlowStatus(flow.id).status == FlowState.UP
+        }
+
+        and: "Flow was rerouted"
+        flowPath != PathHelper.convert(northboundService.getFlowPath(flow.id))
+        Wrappers.wait(WAIT_OFFSET) { northboundService.getFlowStatus(flow.id).status == FlowState.UP }
```
Comment (rtretyak) [inline services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy]: You've already checked this on previous step. Don't understand why to check again

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/BaseSpecification.groovy [hunk 3]
```diff
@@ -46,15 +46,15 @@ class BaseSpecification extends SpringSpecification implements SetupOnce {
     }
 
     @HealthCheck
-    def "Create casual flow to ensure Kilda's up state"() {
+    def "Kilda is UP"() {
```
[the remark this answers is not in the document]
Comment (ylobankov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/BaseSpecification.groovy]: It is just more common and popular test name format.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy [hunk 4]
```diff
@@ -46,6 +47,41 @@ class AutoRerouteSpec extends BaseSpecification {
...
+            possibleFlowPaths.size() > 1
+        } ?: assumeTrue("No suiting switches found", false)
+
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flow.maximumBandwidth = 1000
+        northboundService.addFlow(flow)
+        def flowPath = PathHelper.convert(northboundService.getFlowPath(flow.id))
+        assert Wrappers.wait(WAIT_OFFSET) { northboundService.getFlowStatus(flow.id).status == FlowState.UP }
+
+        when: "Bring port down on the source switch that is involved in the current flow path"
```
[the remark this answers is not in the document]
Comment (ylobankov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy]: Ok, makes sense.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy (hunk 2 above)
Comment (ylobankov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy]: Yes, thanks.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy (hunk 1 above)
Comment (ylobankov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy]: Oops, will fix.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy [hunk 5]
```diff
@@ -46,218 +48,283 @@ class AutoRerouteSpec extends BaseSpecification {
...
-        lockKeeperService.knockoutSwitch(flow.source.datapath)
+        when: "An intermediate switch is disconnected"
+        lockKeeperService.knockoutSwitch(flowPath[1].switchId)
 
-        then: "Flow becomes 'Down'"
-        Wrappers.wait(discoveryTimeout + rerouteDelay + WAIT_OFFSET * 2) {
-            northboundService.getFlowStatus(flow.id).status == FlowState.DOWN
-        }
+        then: "The flow was rerouted after reroute timeout"
+        TimeUnit.SECONDS.sleep(discoveryTimeout + rerouteDelay + WAIT_OFFSET * 2)
```
[the remark this answers is not in the document]
Comment (ylobankov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy]: Done.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy [hunk 6]
```diff
@@ -46,218 +48,283 @@ class AutoRerouteSpec extends BaseSpecification {
...
         }
+    }
+
+    @Unroll
+    def "Flow goes to 'Down' status when #switchType switch is disconnected (#flowType)"() {
+        requireProfiles("virtual")
 
-        when: "The destination switch is disconnected"
-        lockKeeperService.knockoutSwitch(flow.destination.datapath)
+        given: "#flowType.capitalize()"
```
[the remark this answers is not in the document]
Comment (ylobankov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy]: Yes, it works fine.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy [hunk 7]
```diff
@@ -277,36 +344,97 @@ class AutoRerouteSpec extends BaseSpecification {
...
+            } ?: assumeTrue("No suiting switches found", false)
+        } else {
+            isl = topology.getIslsForActiveSwitches().first()
+            if (getAllPaths) {
+                allFlowPaths = db.getPaths(isl.srcSwitch.dpId, isl.dstSwitch.dpId)*.path.sort { it.size() }
+            }
+        }
+
+        def flow = flowHelper.randomFlow(isl.srcSwitch, isl.dstSwitch)
+        getAllPaths ? [flow, allFlowPaths] : flow
```
[the remark this answers is not in the document]
Comment (ylobankov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy]: Yes, it was made for internal usage.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy [hunk 8]
```diff
@@ -46,218 +48,283 @@ class AutoRerouteSpec extends BaseSpecification {
...
-        when: "The source switch is connected back"
-        lockKeeperService.reviveSwitch(flow.source.datapath)
+        PathHelper.convert(northboundService.getFlowPath(flow.id)) != flowPath
 
-        then: "Flow becomes 'Up'"
-        Wrappers.wait(rerouteDelay + discoveryInterval + WAIT_OFFSET) {
-            northboundService.getFlowStatus(flow.id).status == FlowState.UP
+        and: "Connect the intermediate switch back and delete the flow"
+        lockKeeperService.reviveSwitch(flowPath[1].switchId)
+        northboundService.deleteFlow(flow.id)
```
[the remark this answers is not in the document]
Comment (ylobankov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy]: Done.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy [hunk 9]
```diff
@@ -46,218 +48,283 @@ class AutoRerouteSpec extends BaseSpecification {
...
+        then: "The flow goes to 'Up' status"
+        Wrappers.wait(rerouteDelay + discoveryInterval + WAIT_OFFSET) {
+            northboundService.getFlowStatus(flow.id).status == FlowState.UP
+        }
+
+        and: "The flow was rerouted"
+        PathHelper.convert(northboundService.getFlowPath(flow.id)) != flowPath
+
+        and: "Connect switches back involved in the original path and delete the flow"
+        disconnectedSwitches.each { lockKeeperService.reviveSwitch(it.switchId) }
```
[the remark this answers is not in the document]
Comment (ylobankov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy]: Done.

File: services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy [hunk 10]
```diff
@@ -277,36 +344,97 @@ class AutoRerouteSpec extends BaseSpecification {
...
+            } ?: assumeTrue("No suiting switches found", false)
+        } else {
+            (srcSwitch, dstSwitch) = switchPairs.first()
+            if (getAllPaths) {
+                allFlowPaths = db.getPaths(srcSwitch.dpId, dstSwitch.dpId)*.path.sort { it.size() }
+            }
+        }
+
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        getAllPaths ? [flow, allFlowPaths] : flow
```
[the remark this answers is not in the document]
Comment (ylobankov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy]: Sorry, where do you see the duplication?

[the remark this answers is not in the document]
Comment (rtretyak) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy]: The only difference between intermediateSwitchFlow and noIntermediateSwitchFlow is the amount of switches in the preferred path, which can become a method parameter.
```
def getFlowWithPaths(int minAltPaths, int minSwitchesInPath, int maxSwitchesInPath) {
   def switches = topology.getActiveSwitches()
   def possibleFlowPaths
   def switchPair = [switches, switches].combinations().findAll { src, dst -> src != dst }
            .unique { it.sort() }
            .find { src, dst ->
               possibleFlowPaths = db.getPaths(src.dpId, dst.dpId)*.path
               def preferredPath = possibleFlowPaths.min { it.size() }
               possibleFlowPaths.size() > minAltPaths &&  preferredPath.size() <= maxSwitchesInPath
 && preferredPath.size() >= minSwitchesInPath
    }
    switchPair ?: assumeTrue("Such flow cannot be built", false)
    return [flowHelper.randomFlow(switchPair[0], switchPair[1]), possibleFlowPaths]
}
```

Now your methods could be just
```
def intermediateSwitchFlow(boolean ensureAltPathsExist = false, boolean getAllPaths = false) {
    def flowWithPaths = getFlowWithPaths(ensureAltPathsExist ? 1 : 0, 3, Integer.MAX_VALUE)
    return getAllPaths ? flowWithPaths : flowWithPaths[0]
}

def noIntermediateSwitchFlow(boolean ensureAltPathsExist = false, boolean getAllPaths = false) {
    def flowWithPaths = getFlowWithPaths(ensureAltPathsExist ? 1 : 0, 2, 2)
    return getAllPaths ? flowWithPaths : flowWithPaths[0]
}
```

[the remark this answers is not in the document]
Comment (ylobankov) [reply services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/AutoRerouteSpec.groovy]: Thanks, fixed.

**Pull request merged**